### PR TITLE
New channels API

### PR DESCRIPTION
**Pull request by qwwdfsad**
Scope of the PR is TBA

### CODE REVIEW

File: kotlinx-coroutines-core/common/src/channels/Channel.kt [hunk 1]
```diff
@@ -85,7 +85,23 @@ public interface SendChannel<in E> {
      * then it calls `onUndeliveredElement` before throwing an exception.
      * See "Undelivered elements" section in [Channel] documentation for details on handling undelivered elements.
      */
-    public fun offer(element: E): Boolean
+    public fun offer(element: E): Boolean {
+        val result = trySend(element)
+        if (result.isSuccess) return true
+        throw recoverStackTrace(result.exceptionOrNull() ?: return false)
```
Comment (qwwdfsad) [inline kotlinx-coroutines-core/common/src/channels/Channel.kt:91]: Here we have a minor design problem: such implementation of `offer` does not invoke `onUndeliveredElement`.

I've added a workaround 3d44ab93074508c70f9eea67414a3ea0c0213a5c that we can maintain until 1.7 version (where `offer` will become `HIDDEN` function with a default implementation).

File: kotlinx-coroutines-core/common/src/channels/Channel.kt [hunk 2]
```diff
@@ -257,32 +273,35 @@ public interface ReceiveChannel<out E> {
...
-     * This function can be used in [select] invocations with the [onReceiveOrClosed] clause.
+     * This function can be used in [select] invocations with the [onReceiveCatching] clause.
      * Use [poll] to try receiving from this channel without waiting.
-     *
-     * @suppress *This is an internal API, do not use*: Inline classes ABI is not stable yet and
-     *            [KT-27524](https://youtrack.jetbrains.com/issue/KT-27524) needs to be fixed.
      */
-    @InternalCoroutinesApi // until https://youtrack.jetbrains.com/issue/KT-27524 is fixed
-    public suspend fun receiveOrClosed(): ValueOrClosed<E>
+    public suspend fun receiveCatching(): ChannelResult<E>
```
Comment (elizarov) [inline kotlinx-coroutines-core/common/src/channels/Channel.kt:279]: Some leftover in the doc still mentions `receiveOrClosed`

File: kotlinx-coroutines-core/common/src/channels/Channel.kt [hunk 3]
```diff
@@ -321,104 +340,104 @@ public interface ReceiveChannel<out E> {
...
- * :todo: Do not make it public before resolving todos in the code of this class.
+ * The successful result represents a successful operation with a value of type [T], for example,
+ * the result of [Channel.receiveCatching] operation or a successfully sent element as a result of [Channel.trySend].
  *
- * @suppress *This is an internal API, do not use*: Inline classes ABI is not stable yet and
- *            [KT-27524](https://youtrack.jetbrains.com/issue/KT-27524) needs to be fixed.
+ * The failed result represents a failed operation attempt to a channel, but it doesn't necessary indicate that the channel is failed.
+ * E.g. when the channel is full, [Channel.trySend] returns failed result, but the channel itself is not in the failed state.
+ *
+ * The closed result represents an operation attempt to a closed channel and also implies that the operation was failed.
```
Comment (elizarov) [inline kotlinx-coroutines-core/common/src/channels/Channel.kt:353]: was failed -> has failed.

**Review by elizarov — COMMENTED** [review]
API looks good so far.

**Pull request merged**